src/tester.zig

```diff
@@ -0,0 +1,184 @@
+        );
+    }
+}
+
+/// Test that `uncompressed_values` are within `error_bound` according to `within_error_bound`
+/// after it has been compressed and decompressed using `method`. The libraries public interface is
+/// used to make it simpler refactor the libraries internals.
+pub fn testCompressAndDecompress(
+    uncompressed_values: []const f64,
+    allocator: Allocator,
+    method: Method,
+    error_bound: f32,
+    within_error_bound: fn (
+        uncompressed_values: []const f64,
+        decompressed_values: []const f64,
+        error_bound: f32,
+    ) bool,
+) !void {
+    const compressed_values = try tersets.compress(
+        uncompressed_values,
+        allocator,
+        method,
+        error_bound,
+    );
+    defer compressed_values.deinit();
+    const decompressed_values = try tersets.decompress(compressed_values.items, allocator);
+    defer decompressed_values.deinit();
+
+    try testing.expectEqual(uncompressed_values.len, decompressed_values.items.len);
+    try testing.expect(within_error_bound(
+        uncompressed_values,
+        decompressed_values.items,
+        error_bound,
+    ));
+}
+
+// Replace each normal value in `uncompressed_values` with a positive infinity, negative infinity,
+// or NaN with the passed probability. The non-normal values are written to `uncompressed_values`
+// in the previously listed order, thus a positive infinity maybe overwritten by a negative infinity
+// and so on. The probabilities are asserted to be between zero and one.
+pub fn replaceNormalValues(
+    uncompressed_values: *ArrayList(f64),
+    positive_infinity_probability: f32,
+    negative_infinity_probability: f32,
+    not_a_number_probability: f32,
+    random: Random,
+) void {
+    debug.assert(0 <= positive_infinity_probability and positive_infinity_probability <= 1);
+    debug.assert(0 <= negative_infinity_probability and negative_infinity_probability <= 1);
+    debug.assert(0 <= not_a_number_probability and not_a_number_probability <= 1);
+
+    for (0..uncompressed_values.items.len) |index| {
+        if (random.float(f32) < positive_infinity_probability) {
+            uncompressed_values.items[index] = math.inf(f64);
+        }
+
+        if (random.float(f32) < negative_infinity_probability) {
+            uncompressed_values.items[index] = -math.inf(f64);
+        }
+
+        if (random.float(f32) < not_a_number_probability) {
+            uncompressed_values.items[index] = math.nan(f64);
+        }
+    }
+}
+
+/// Generate `number_of_values` of random values for use in testing using `random` and add them to
+/// `uncompressed_values`.
+pub fn generateRandomValues(uncompressed_values: *ArrayList(f64), random: Random) !void {
+    for (0..number_of_values) |_| {
+        // math.floatMax(f64) is not used as it larger than -math.floatMin(f64).
+        try uncompressed_values.append(math.floatMin(f64) +
+            (-math.floatMin(f64) - math.floatMin(f64)) * rand.float(random, f64));
```

Comment:
> Would not `(-math.floatMin(f64) - math.floatMin(f64))` generate 0 or almost 0?

Yes, but I tried to an make it clear how the function generates a range of values as Zig's [`std.Random` ](https://ziglang.org/documentation/master/std/#std.Random)only supports generating floats in the range [0, 1). Since this function is only for generating test values and not anything related to cryptography I looked into how other programming languages generates random float values and Python had a nice simple method. Like Zig, Python's [`random.random()`](https://docs.python.org/3/library/random.html#random.random) also only generates values in the range [0, 1) but they also provide the method[ `random.uniform(a, b)`](https://docs.python.org/3/library/random.html#random.uniform)  for generating values in from `a` to `b`. You can see the implementation of [`random.uniform(a, b)` here](https://github.com/python/cpython/blob/main/Lib/random.py#L498). Since the function is intended to be used to generate random normal `f64` values for testing I only tested it by generating a few sets of values and checking that they looked reasonable. So if there are any issues related to the generated values we should probably use another method, unfortunately I am not aware of other methods for generating random normal `f64` values.

> Also, this Eq. can generate numbers that are less than `math.floatMin(f64)` is that the expected behavior?

How is this possible when [`math.floatMin(f64)`](https://ziglang.org/documentation/master/std/#std.math.floatMin) returns the smallest normal number representable in `f64` according to its documentation?

Reply:
Well, I was doing several tests of this function and why does not seem to be working. I think the problem is that `math.floatMin(f64)` is equal to `2.2250738585072014e-308` which is a very small positive number `0.00...`, subtracting against its inverse you obtain `-4.450147717014403e-308` which is a very small negative number `-0.00...`. Perhaps a better solution is to generate a random 64 bits number and then cast it to a floating point like this:
`const random_u64 = std.rand.int(random, u64);`
`const random_f64 = @as(f64, @bitCast(random_u64));`
I did some tests with this and seems to be working quite well. Perhaps you can change the current function to do something like it or find a solution for the `max-min` range.

Reply:
> Well, I was doing several tests of this function and why does not seem to be working. I think the problem is that `math.floatMin(f64)` is equal to `2.2250738585072014e-308` which is a very small positive number `0.00...`, subtracting against its inverse you obtain `-4.450147717014403e-308` which is a very small negative number `-0.00...`. Perhaps a better solution is to generate a random 64 bits number and then cast it to a floating point like this: `const random_u64 = std.rand.int(random, u64);` `const random_f64 = @as(f64, @bitCast(random_u64));` I did some tests with this and seems to be working quite well. Perhaps you can change the current function to do something like it or find a solution for the `max-min` range.

Consider this!